guest/RATIONALE.md

```diff
@@ -47,7 +47,38 @@ mean amplifying that overhead to a point where second or longer latency would
 be possible. For reasons like this, we dismissed Go 1.21 for TinyGo, and will
 revisit when Go 1.22 introduces [`//go:wasmexport`][3].
 
+## Why do we use nottinygc?
+
+Kubernetes Scheduler plugins need to access very large model data, specifically
+node and pod data. WebAssembly has a sandbox model, so the memory of a plugin
+is not the same as host scheduler process. This means the node and pod are not
+passed by reference. In fact they cannot be copied by value either. Instead,
+they must be marshaled to wire format and unmarshalled within the WebAssembly
+guest.
+
+Unmarshalling protobuf messages with default tooling creates a lot of garbage.
+As WebAssembly has only one thread, garbage collection is inlined. TinyGo's
+compiler optimized for code size not speed. In the default configuration, we
+found inlined GC overhead to be over half the latency of a plugin execution.
+
+[nottinygc][4] is an alternate garbage collection implementation for TinyGo.
+This optimizes for performance instead of bytecode size. Swapping this added
+110KB to the base size of the wasm guest, but resulted in 48pct drop in
+execution time of a simple plugin, in a scale of hundreds of microseconds.
+
+nottinygc has tradeoffs besides size. One is that it isn't built-in to TinyGo.
```

Comment:
you are going to yarn ball this into a bit PR, but that's ok ;)

scheduler=none iirc is here (already before nottinygc) because some inherited understanding about function exports besides _start aka main. IIUC you can't really use goroutines anywhere except main, so `scheduler=none` prevents accidental use. If you remember exact, I can jot down a better one.

Reply:
Oh yeah I forgot about that - so yeah it's not really related to this PR so no need to yarn ball ;)

Reply:
I'll leave it out for now, just because anything we add needs digging to polish it, and others to review. I don't have a good blurb for this and can work on it on any other project and paste a refined one here. so many projects add `scheduler=none` it should be easy enough to get some polished answer available.

Reply:
wait we already have something.. lemme fix it https://wazero.io/languages/tinygo/#concurrency

Reply:
ok PTAL as luckily the text work was already done by wazero!